src/zerocoin_v3.cpp

```diff
@@ -460,8 +463,41 @@ bool CheckSigmaTransaction(
     return true;
 }
 
-void DisconnectTipSigma(CBlock & /*block*/, CBlockIndex *pindexDelete) {
+void RemoveSigmaSpendsReferencingBlock(CBlockIndex* blockIndex) {
+    std:vector<CTransaction> txn_to_remove;
+    for (CTxMemPool::txiter mi = mempool.mapTx.begin(); mi != mempool.mapTx.end(); ++mi) {
```

Comment:
We need to lock before accessing `mempool`

Reply:
We are locking upper, for example in DisconnectBlocks() . Why do we need to lock here? May be better to call AssertLockHeld() here?

Reply:
theoretically it's ok to have it run only under `cs_main` lock which is acquired before the call. But adding one more lock won't hurt

Reply:
also have a minor question to @martun : what about dandelion pool? don't we need to delete txs from there as well?

Reply:
Having them left in stempool does not hurt, but I will change the code to remove also from there. Will also add additional locks, they will not hurt.